Sources/SpeziFHIR/FHIRStore.swift

```diff
@@ -19,149 +17,156 @@ import Spezi
         }
     }
-    
-    /// Loads resources from a given FHIR `Bundle`.
+
+    /// Loads resources from a given FHIR `Bundle` into the ``FHIRStore``.
     ///
     /// - Parameter bundle: The FHIR `Bundle` containing resources to be loaded.
-    public func load(bundle: Bundle) {
+    public func load(bundle: sending Bundle) async {
         let resourceProxies = bundle.entry?.compactMap { $0.resource } ?? []
-        
+        var resources: [FHIRResource] = []
+
         for resourceProxy in resourceProxies {
-            insert(resource: FHIRResource(resource: resourceProxy.get(), displayName: resourceProxy.displayName))
+            if Task.isCancelled {
+                return
+            }
+
+            resources.append(
+                FHIRResource(
+                    resource: resourceProxy.get(),
+                    displayName: resourceProxy.displayName
+                )
+            )
+        }
+
+        if Task.isCancelled {
+            return
         }
+
+        await insert(resources: resources)
+    }
+
+    /// Removes a FHIR resource from the ``FHIRStore``.
+    ///
+    /// - Parameter resource: The `FHIRResource` identifier to be inserted.
+    @MainActor
+    public func remove(resource resourceId: FHIRResource.ID) {
+        guard let resource = _resources.first(where: { $0.id == resourceId }) else {
+            return
+        }
+
+        _$observationRegistrar.willSet(self, keyPath: resource.category.storeKeyPath)
+
+        _resources.removeAll { $0.id == resourceId }
+
+        _$observationRegistrar.didSet(self, keyPath: resource.category.storeKeyPath)
     }
-    
-    /// Removes all resources from the store.
+
+    /// Removes all resources from the ``FHIRStore``.
+    @MainActor
     public func removeAllResources() {
-        lock.withLock {
-            // Not really ideal but seems to be a path to ensure that all observables are called.
-            _$observationRegistrar.willSet(self, keyPath: \.allergyIntolerances)
-            _$observationRegistrar.willSet(self, keyPath: \.conditions)
-            _$observationRegistrar.willSet(self, keyPath: \.diagnostics)
-            _$observationRegistrar.willSet(self, keyPath: \.encounters)
-            _$observationRegistrar.willSet(self, keyPath: \.immunizations)
-            _$observationRegistrar.willSet(self, keyPath: \.medications)
-            _$observationRegistrar.willSet(self, keyPath: \.observations)
-            _$observationRegistrar.willSet(self, keyPath: \.otherResources)
-            _$observationRegistrar.willSet(self, keyPath: \.procedures)
-            _resources = []
-            _$observationRegistrar.didSet(self, keyPath: \.allergyIntolerances)
-            _$observationRegistrar.didSet(self, keyPath: \.conditions)
-            _$observationRegistrar.didSet(self, keyPath: \.diagnostics)
-            _$observationRegistrar.didSet(self, keyPath: \.encounters)
-            _$observationRegistrar.didSet(self, keyPath: \.immunizations)
```

Comment:
Something that this PR is changing is that previously the observation pattern is much more granular. If you were just interested in a certain category, the view would just update if an element in this category was changed. Might this impact performance for our use case?

Reply:
Not sure what you mean by that.
The `insert(resource: FHIRResource)` just triggers the observation registrar of the specific resource type, the `insert<T: Collection>(resources: T)` triggers the registrar for all resources types that were inserted.
Same thing goes for the removal of resources.